File: elixir/bob/lib/bob.ex
defmodule Bob do
  def hey(input) do
    cond do
      input |> String.trim() == "" ->
        "Fine. Be that way!"

      input == input |> String.upcase() and input |> String.last() == "?" and
          input != input |> String.downcase() ->
        "Calm down, I know what I'm doing!"

      input == input |> String.upcase() and input != input |> String.downcase() ->
        "Whoa, chill out!"

      input |> String.last() == "?" ->
        "Sure."

      true ->
        "Whatever."
    end
  end
end

File: clojure/bob/src/bob.clj
(ns bob
  (:require [clojure.string :as str]))

(defn- is-empty? [s]
  (= (str/trim s) ""))

(defn- is-question? [s]
  (= (last s) \?))

(defn- is-upcase? [s]
  (and (not-any? #(Character/isLowerCase %) s) (some #(Character/isUpperCase %) s)))
  
(defn response-for [s]
  (cond
    (is-empty? s) "Fine. Be that way!"
    (and (is-question? s) (is-upcase? s)) "Calm down, I know what I'm doing!"
    (is-upcase? s) "Whoa, chill out!"
    (is-question? s) "Sure."
    :else "Whatever."))

